### ISSUE TITLE
Sampled softmax loss stops gradients on sampled classes

### ISSUE THREAD

**Issue by markostam** [tensorflow/tensorflow#16552]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
Binary
- **TensorFlow version (use command below)**:
1.3
- **Python version**: 
2.7
- **Bazel version (if compiling from source)**:
N/A
- **GCC/Compiler version (if compiling from source)**:
N/A
- **CUDA/cuDNN version**:
CUDA 8 / cuDNN 6
- **GPU model and memory**:
4 x TITAN X (Pascal)

### Describe the problem

The backbone of TensorFlow's sampled loss functions `nce_loss` and `sampled_softmax_loss` is a helper function called `_compute_sampled_logits`https://github.com/tensorflow/tensorflow/blob/8fb12848d3a81a010714a4612ffd735106ea83d8/tensorflow/python/ops/nn_impl.py#L961-L1139.

`_compute_sampled_logits` takes as input:

- weights and biases of the final layer,
- the output labels
- the inputs to the final layer inputs
- the sampled values of the output layer
- a few other things

and returns the logits and labels of only the requested sampled labels.

One of the first ops executed is https://github.com/tensorflow/tensorflow/blob/8fb12848d3a81a010714a4612ffd735106ea83d8/tensorflow/python/ops/nn_impl.py#L1046-L1047 

This line seems like it is stopping gradients flowing back through the sampled values if i'm reading it correctly.

Shouldn't the gradients be stopped from flowing back through the _non-sampled values_ as opposed to the _sampled values_? Why are gradients being stopped at the sampled values?

**Comment by drpngx** [comment]
@gouwsmeister could you take a look?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by gouwsmeister** [comment]
`sampled` is only a list of integer ids of the sampled positive classes. The actual weights are selected/indexed further down. Applying stop_gradient on the ids does not prevent the actual weights to still be updated.

**Comment by drpngx** [comment]
Closing since this seems to be correct behavior.
